File: services/core/internal/bot/announce.go
package bot

import (
	"context"
	"errors"
	"log/slog"
	"sync"
	"time"

	"github.com/bwmarrin/discordgo"
	"github.com/minhhdtr/xsmb-discord-bot/internal/domain"
)

// Poster sends one embed to one channel.
type Poster func(ctx context.Context, channelID string, embed *discordgo.MessageEmbed) error

// Announcer posts each day's result once 18:35 has passed and the page has
// all 27 numbers.
type Announcer struct {
	core     Core
	now      func() time.Time
	post     Poster
	log      *slog.Logger
	window   time.Duration // how long to keep polling after the draw time
	interval time.Duration // pause between polls
	gap      time.Duration // pause between sends within one worker
	workers  int           // channels posted to at once
}

// NewAnnouncer builds the scheduler. window bounds how long a stalled draw is
// chased before the run is abandoned until tomorrow.
func NewAnnouncer(core Core, clock func() time.Time, post Poster, log *slog.Logger) *Announcer {
	if log == nil {
		log = slog.Default()
	}
	if clock == nil {
		clock = func() time.Time { return time.Now().In(domain.Location()) }
	}
	return &Announcer{
		core: core, now: clock, post: post, log: log,
		window:   45 * time.Minute,
		interval: 20 * time.Second,
		gap:      150 * time.Millisecond,
		workers:  6,
	}
}

// SetPollInterval overrides the retry interval. For tests.
func (a *Announcer) SetPollInterval(d time.Duration) { a.interval = d }

// SetFanOut overrides the worker count and the pause between sends.
func (a *Announcer) SetFanOut(workers int, gap time.Duration) {
	if workers > 0 {
		a.workers = workers
	}
	a.gap = gap
}

// Run blocks until ctx ends, announcing once a day. On startup it catches up
// on today if 18:35 has already passed; the DB claim stops duplicates.
func (a *Announcer) Run(ctx context.Context) {
	if now := a.now(); !now.Before(domain.CompletionTime(now)) {
		a.RunFor(ctx, domain.DayOf(now))
	}
	for {
		wait := a.untilNext()
		a.log.Info("announcer waiting", "next_run", a.now().Add(wait).Format(time.RFC3339))
		select {
		case <-ctx.Done():
			a.log.Info("announcer stopped")
			return
		case <-time.After(wait):
		}
		a.RunFor(ctx, domain.DayOf(a.now()))
	}
}

// untilNext is the delay to the next 18:35. Recomputed each cycle so clock
// drift can't accumulate.
func (a *Announcer) untilNext() time.Duration {
	now := a.now()
	next := domain.CompletionTime(now)
	if !now.Before(next) {
		next = domain.CompletionTime(now.AddDate(0, 0, 1))
	}
	return next.Sub(now)
}

// RunFor waits for day's result, then posts it to every subscribed channel.
func (a *Announcer) RunFor(ctx context.Context, day time.Time) {
	subs, err := a.core.Subscriptions(ctx)
	if err != nil {
		a.log.Error("cannot list subscriptions", "error", err)
		return
	}
	if len(subs) == 0 {
		a.log.Info("no subscribed channels, skipping announcement", "date", domain.FormatISO(day))
		return
	}

	waitCtx, cancel := context.WithTimeout(ctx, a.window)
	defer cancel()

	draw, err := awaitComplete(waitCtx, a.core, day, a.interval)
	if err != nil {
		if errors.Is(err, domain.ErrNoResult) {
			a.log.Info("no draw to announce", "date", domain.FormatISO(day))
			return
		}
		a.log.Error("giving up on announcement", "date", domain.FormatISO(day), "error", err)
		return
	}

	embed := DrawEmbed(draw, true)
	sent := a.fanOut(ctx, day, subs, embed)

	a.log.Info("announcement done", "date", domain.FormatISO(day), "channels", sent)
}

// fanOut posts to all subscribed channels a few at a time. Discord rate
// limits per channel, so different channels can be written to in parallel.
func (a *Announcer) fanOut(ctx context.Context, day time.Time,
	subs []domain.Subscription, embed *discordgo.MessageEmbed) int {

	workers := a.workers
	if workers < 1 {
		workers = 1
	}
	if workers > len(subs) {
		workers = len(subs)
	}

	queue := make(chan domain.Subscription)
	go func() {
		defer close(queue)
		for _, sub := range subs {
			select {
			case queue <- sub:
			case <-ctx.Done():
				return
			}
		}
	}()

	var (
		mu   sync.Mutex
		sent int
		wg   sync.WaitGroup
	)
	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for sub := range queue {
				if ctx.Err() != nil {
					return
				}
				if a.postTo(ctx, day, sub.ChannelID, embed) {
					mu.Lock()
					sent++
					mu.Unlock()
				}
				if a.gap > 0 {
					select {
					case <-time.After(a.gap):
					case <-ctx.Done():
						return
					}
				}
			}
		}()
	}
	wg.Wait()
	return sent
}

// postTo claims one channel and sends, reporting whether a message went out.
func (a *Announcer) postTo(ctx context.Context, day time.Time,
	channelID string, embed *discordgo.MessageEmbed) bool {

	won, err := a.core.ClaimAnnouncement(ctx, day, channelID)
	if err != nil {
		a.log.Error("cannot claim announcement", "channel", channelID, "error", err)
		return false
	}
	if !won {
		return false // already posted, by an earlier run or another instance
	}
	if err := a.post(ctx, channelID, embed); err != nil {
		a.log.Error("cannot post announcement", "channel", channelID, "error", err)
		// Give the claim back so the next run retries.
		if err := a.core.ReleaseAnnouncement(ctx, day, channelID); err != nil {
			a.log.Error("cannot release claim", "channel", channelID, "error", err)
		}
		return false
	}
	return true
}

File: services/bot/src/commands/announce.ts
import type { APIEmbed } from "discord.js";
import { Core, CoreError } from "../core/client.js";
import { drawEmbed } from "../render/stats.js";
import { now, ZONE } from "../render/dates.js";
import { DateTime } from "luxon";

/** Where a finished post goes. Keeping this a function rather than a client
 * means the schedule can be tested without a gateway connection. */
export type Poster = (channelId: string, embed: APIEmbed) => Promise<void>;

export interface AnnouncerOptions {
  /** How long to keep asking before giving up until tomorrow. */
  windowMs?: number;
  /** Gap between attempts. */
  intervalMs?: number;
  /** Pause between sends, so a fan-out to many channels does not burst. */
  gapMs?: number;
}

/**
 * Posts the day's result to every subscribed channel, once.
 *
 * It does not crawl. Core's own ingest fills the archive on its own schedule
 * whether or not anybody is subscribed — that separation is what fixed the
 * 04/09 bug, where turning off the last subscription silently stopped the
 * archive from updating. This only has to notice.
 *
 * Asking rather than being pushed also keeps the dependency pointing one way:
 * the bot knows where core is, core knows nothing about the bot. A second
 * client can be added without core changing at all.
 */
export class Announcer {
  readonly #window: number;
  readonly #interval: number;
  readonly #gap: number;

  constructor(
    private readonly core: Core,
    private readonly post: Poster,
    options: AnnouncerOptions = {},
  ) {
    this.#window = options.windowMs ?? 45 * 60_000;
    this.#interval = options.intervalMs ?? 20_000;
    this.#gap = options.gapMs ?? 150;
  }

  /** Runs until `signal` aborts. */
  async run(signal: AbortSignal): Promise<void> {
    while (!signal.aborted) {
      const wait = msUntilNextRun();
      console.log(`announcer waiting ${Math.round(wait / 60_000)} minutes`);
      if (await sleep(wait, signal)) return;
      if (signal.aborted) return;

      await this.runFor(now().toISODate() ?? "", signal);
    }
  }

  /**
   * Chases one day's result and posts it.
   *
   * The claim is taken before the post and given back if the post fails, so a
   * restart at 18:50 does not put the same result out twice and a Discord
   * hiccup does not lose the day.
   */
  async runFor(date: string, signal?: AbortSignal): Promise<void> {
    const draw = await this.#await(date, signal);
    if (!draw) return;

    let subscriptions;
    try {
      subscriptions = await this.core.subscriptions();
    } catch (error) {
      console.error("cannot read subscriptions:", error);
      return;
    }
    if (subscriptions.length === 0) {
      console.log(`archived ${date}, no channels to announce to`);
      return;
    }

    const embed = drawEmbed(draw, true);
    for (const subscription of subscriptions) {
      if (signal?.aborted) return;
      await this.#postOnce(date, subscription.channel_id, embed);
      await sleep(this.#gap);
    }
  }

  async #postOnce(date: string, channelId: string, embed: APIEmbed): Promise<void> {
    let won: boolean;
    try {
      won = await this.core.claimAnnouncement(date, channelId);
    } catch (error) {
      console.error(`cannot claim ${date} for ${channelId}:`, error);
      return;
    }
    if (!won) return; // Someone already posted it. Staying quiet is the point.

    try {
      await this.post(channelId, embed);
    } catch (error) {
      console.error(`cannot post to ${channelId}:`, error);
      // Give the claim back so the next run tries again rather than the day
      // being lost to a transient failure.
      await this.core.releaseAnnouncement(date, channelId).catch(() => {});
      return;
    }

    // Sent. Ending the lease here rather than at claim time is what makes a
    // crash recoverable: a claim left unsent expires and can be taken over,
    // while this one never will be.
    await this.core.markAnnounced(date, channelId).catch((error: unknown) => {
      // The message is out. Failing to record that only risks a duplicate
      // after a crash inside the lease window, which is the lesser problem.
      console.error(`posted to ${channelId} but could not mark it sent:`, error);
    });
  }

  /** Asks core for the day until it has it, or the window closes. */
  async #await(date: string, signal?: AbortSignal) {
    const deadline = Date.now() + this.#window;

    while (Date.now() < deadline) {
      if (signal?.aborted) return null;
      try {
        return await this.core.draw(date);
      } catch (error) {
        if (worthRetrying(error)) {
          // Two different reasons to wait, both temporary. `not_yet` is the
          // ordinary one: the source publishes tier by tier, so the first
          // several minutes after the mark look like this. A timeout or a 5xx
          // is core being restarted or briefly unreachable — which used to
          // lose the whole day, because any error that was not `not_yet` ended
          // the attempt.
          if (await sleep(this.#interval, signal)) return null;
          continue;
        }
        console.error(`cannot get ${date}, giving up:`, error);
        return null;
      }
    }
    console.warn(`gave up on ${date} after ${Math.round(this.#window / 60_000)} minutes`);
    return null;
  }
}

/**
 * Whether an error is worth waiting through.
 *
 * The split is between "not there yet" and "will never be there". A day that
 * is settled and empty, or outside the archive, cannot be fixed by asking
 * again; a timeout, a 5xx, or an unreachable core all can. Getting this wrong
 * in the safe direction only costs requests inside a window that closes on its
 * own after forty-five minutes.
 */
export function worthRetrying(error: unknown): boolean {
  if (!(error instanceof CoreError)) {
    // fetch itself failed: DNS, connection refused, the socket dropped.
    return true;
  }
  if (error.notYet) return true;
  if (error.noDraw || error.code === "out_of_range" || error.code === "bad_request") {
    return false;
  }
  // 0 is a client-side failure; 5xx is core's problem and usually brief.
  return error.status === 0 || error.status >= 500;
}

/**
 * Milliseconds until the next 18:35 in Hanoi.
 *
 * Computed from the clock each time rather than by adding 24 hours, so a
 * restart at 18:40 waits until tomorrow instead of firing immediately, and a
 * process left running for weeks does not drift.
 */
export function msUntilNextRun(from: DateTime = now()): number {
  const at = from.setZone(ZONE);
  let next = at.set({ hour: 18, minute: 35, second: 0, millisecond: 0 });
  if (next <= at) next = next.plus({ days: 1 });
  return next.diff(at).toMillis();
}

/** Resolves true if the wait was cut short by the signal. */
function sleep(ms: number, signal?: AbortSignal): Promise<boolean> {
  return new Promise((resolve) => {
    if (signal?.aborted) return resolve(true);
    const timer = setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve(false);
    }, ms);
    const onAbort = () => {
      clearTimeout(timer);
      resolve(true);
    };
    signal?.addEventListener("abort", onAbort, { once: true });
  });
}
